File: session/03_visitor/clojure/subsumed_visitor.clj
(defrecord Cylinder [baseRadius, height])
(defrecord Sphere [radius])

(defmulti volume class)

(defmethod volume Cylinder [{:keys [baseRadius height]}]
  (* (. Math PI) (Math/pow baseRadius 2) height))
  
(defmethod volume Sphere [{:keys [radius]}] 
  (* (double (/ 4 3)) (. Math PI) (Math/pow radius 3)))

(defmethod volume :default [s] nil)


(defmulti surface-area class)

(defmethod surface-area Cylinder [{:keys [baseRadius height]}]
	(let [base-area (* (. Math PI) (Math/pow baseRadius 2))
	    base-circumference (* 2 (. Math PI) baseRadius)]
	    (+ (* 2 base-area) (* base-circumference height))))
  
(defmethod surface-area Sphere [{:keys [radius]}] 
  (* 4 (. Math PI) (Math/pow radius 2)))

(defmethod surface-area :default [s] nil)

(def cylinder (Cylinder. 10 10))
(def sphere (Sphere. 10))

(println (volume cylinder)) ; 3141.59
(println (volume sphere)) ; 4188.79
(println (volume 2)) ;nil

(println (surface-area cylinder)) ;1256.63
(println (surface-area sphere)) ;1256.63
(println (surface-area 2)) ;nil

File: session/03_visitor/scala/subsumed_visitor.scala
trait Shape3d {
  def surfaceArea(): Double
  def volume(): Double
}

case class Cylinder(baseRadius: Double, height: Double) extends Shape3d {
  val baseArea = Math.PI * Math.pow(baseRadius, 2)
  val baseCircumference = 2 * Math.PI * baseRadius
  def surfaceArea(): Double = 2 * baseArea + baseCircumference * height
  def volume(): Double = baseArea * height
}

case class Sphere(radius: Double) extends Shape3d {
  def surfaceArea(): Double = 4 * Math.PI * Math.pow(radius, 2)
  def volume(): Double = 4.0 / 3.0 * Math.PI * Math.pow(radius, 3)
}

case class CompositeShape(shapes: List[Shape3d]) extends Shape3d {
  def surfaceArea(): Double = shapes.map(_.surfaceArea()).sum
  def volume(): Double = shapes.map(_.volume()).sum
}

def renderOpenGL(s: Shape3d): Unit = s match {
  case Cylinder(baseRadius, height) => println("OpenGL: rendering cylinder")
  case Sphere(radius) => println("OpenGL: rendering sphere")
  case CompositeShape(shapes) => shapes foreach renderOpenGL
}

def renderDirect3D(s: Shape3d): Unit = s match {
  case Cylinder(baseRadius, height) => println("Direct3D: rendering cylinder")
  case Sphere(radius) => println("Direct3D: rendering sphere")
  case CompositeShape(shapes) => shapes foreach renderDirect3D
}

val sphere = Sphere(10)
val cylinder = Cylinder(10, 10)
val composite = CompositeShape(List(cylinder, sphere))
println(composite.surfaceArea())
println(composite.volume())

renderOpenGL(composite)
renderDirect3D(composite)
